go-runtime/ftl/ftltest/fake.go

```diff
@@ -4,33 +4,97 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"math/rand"
 	"reflect"
 	"strings"
+	"sync"
+	"time"
 
+	"github.com/TBD54566975/ftl/backend/schema"
 	"github.com/TBD54566975/ftl/common/configuration"
+	"github.com/TBD54566975/ftl/go-runtime/ftl"
 	"github.com/TBD54566975/ftl/go-runtime/internal"
+	"github.com/TBD54566975/ftl/internal/log"
+	"github.com/TBD54566975/ftl/internal/slices"
+	"github.com/alecthomas/types/optional"
+	"github.com/alecthomas/types/pubsub"
 )
 
+// pubSubEvent is a sum type for all events that can be published to the pubsub system.
+// not to be confused with an event that gets published to a topic
+//
+//sumtype:decl
+type pubSubEvent interface {
+	// cronJobEvent is a marker to ensure that all events implement the interface.
+	pubSubEvent()
+}
+
+// publishEvent holds an event to be published to a topic
+type publishEvent struct {
+	topic   *schema.Ref
+	content any
+}
+
+func (publishEvent) pubSubEvent() {}
+
+// subscriptionDidConsumeEvent indicates that a call to a subscriber has completed
+type subscriptionDidConsumeEvent struct {
+	subscription string
+	err          error
+}
+
+func (subscriptionDidConsumeEvent) pubSubEvent() {}
+
+type subscription struct {
+	name        string
+	topic       *schema.Ref
+	cursor      optional.Option[int]
+	isExecuting bool
+	errors      map[int]error
+}
+
+type subscriber func(context.Context, any) error
+
 type fakeFTL struct {
-	fsm           *fakeFSMManager
+	fsm *fakeFSMManager
+
 	mockMaps      map[uintptr]mapImpl
 	allowMapCalls bool
 	configValues  map[string][]byte
 	secretValues  map[string][]byte
+
+	// all pubsub events are processed through globalTopic
+	globalTopic *pubsub.Topic[pubSubEvent]
+	// publishWaitGroup can be used to wait for all events to be published
+	publishWaitGroup sync.WaitGroup
+
+	// pubSubLock required to access [topics, subscriptions, subscribers]
+	pubSubLock    sync.Mutex
+	topics        map[schema.RefKey][]any
+	subscriptions map[string]*subscription
+	subscribers   map[string][]subscriber
 }
 
 // mapImpl is a function that takes an object and returns an object of a potentially different
 // type but is not constrained by input/output type like ftl.Map.
 type mapImpl func(context.Context) (any, error)
 
-func newFakeFTL() *fakeFTL {
-	return &fakeFTL{
+func newFakeFTL(ctx context.Context) *fakeFTL {
+	fake := &fakeFTL{
 		fsm:           newFakeFSMManager(),
 		mockMaps:      map[uintptr]mapImpl{},
 		allowMapCalls: false,
 		configValues:  map[string][]byte{},
 		secretValues:  map[string][]byte{},
+		globalTopic:   pubsub.New[pubSubEvent](),
```

Comment:
Let's pull all the pubsub code out into a separate struct and file, and forward to it from fakeFTL.